internal/worker/apiremotecaller/remote.go

```diff
@@ -0,0 +1,312 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package apiremotecaller
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/juju/clock"
+	"github.com/juju/errors"
+	"github.com/juju/names/v5"
+	"github.com/juju/retry"
+	"github.com/juju/worker/v4"
+	"gopkg.in/tomb.v2"
+
+	"github.com/juju/juju/api"
+	"github.com/juju/juju/core/logger"
+)
+
+// RemoteConnection is an interface that represents a connection to a remote
+// API server.
+type RemoteConnection interface {
+	// Connection returns the connection to the remote API server.
+	Connection() api.Connection
+
+	// Tag returns the tag of the remote API server.
+	Tag() names.Tag
+}
+
+// RemoteServer represents the public interface of the worker
+// responsible modeling the remote API server.
+type RemoteServer interface {
+	worker.Worker
+	RemoteConnection
+	UpdateAddresses(addresses []string)
+}
+
+// RemoteServerConfig defines all the attributes that are needed for a
+// RemoteServer.
+type RemoteServerConfig struct {
+	Clock  clock.Clock
+	Logger logger.Logger
+
+	Target names.Tag
+
+	// APIInfo is initially populated with the addresses of the target machine.
+	APIInfo *api.Info
+
+	// APIOpener is a function that will open a connection to the target API
+	// server.
+	APIOpener api.OpenFunc
+}
+
+// remoteServer is a worker that provides addresses for the target API server.
+type remoteServer struct {
+	internalStates chan string
+	tomb           tomb.Tomb
+
+	target names.Tag
```

Comment:
Let's just use controller ID here. I know the pub/sub message hands around a tag, but we should _decode_ it at the earliest opportunity and base logic on IDs.